rhizome/lib/common.rb

```diff
@@ -18,8 +18,28 @@ def to_s
   end
 end
 
-def r(commandline)
-  stdout, stderr, status = Open3.capture3(commandline)
+class FsyncFail < RuntimeError
+end
+
+def r(commandline, stdin: "")
+  stdout, stderr, status = Open3.capture3(commandline, stdin_data: stdin)
   fail CommandFail.new("command failed: " + commandline, stdout, stderr) unless status.success?
   stdout
 end
+
+def fsync_or_fail(f)
+  # Throw a custom exception type, so we can page if fsync error happens.
+  # Note that IO::fsync raises an exception on error based on its source
+  # in the docs: https://ruby-doc.org/core-2.4.2/IO.html#method-i-fsync
+
+  f.fsync
```

Comment:
```suggestion
  # Throw a custom exception type inheriting directly from Exception,
  # unlikely to be accidentally rescued as to better halt the program
  # in event of fsync errors.
  #
  # The ultimate goal of fsync errors is to page.  Halting progress is
  # one roundabout but easy way of doing that.
  #
  # Note that IO::fsync raises an exception on error based on its source
  # in the docs: https://ruby-doc.org/core-2.4.2/IO.html#method-i-fsync
  f.fsync
```

This comment will probably make more sense once @byucesoy writes the deadline implementation.